### PR TITLE
[ALLUXIO-1856] When authorization is not turned on, changed the default permission to full permission

### PR DESCRIPTION
**Pull request by apc999**
Forked from https://github.com/Alluxio/alluxio/pull/2971/files#diff-2f2622ebb203e8d12cf5ef8930205e2e

A few caveats for this PR:
- It turns out the major complexity of this PR is to figure out where to apply UMask logic. After a few iterations, I finally put the UMask logic into the `createPath` method in class `InodeTree`. The reason behind is that UMask only matters for newly created files/directories, and `InodeTree.createPath` is the critical path for those operations.
- Like Linux, new files created in Alluxio have default permission bits 644 and directories  have 755. This is achieved by two different ApplyUMask functions, one for file and one for dir in this PR.

### CODE REVIEW

**Comment by jsimsa** [comment]
LGTM

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 1]
```diff
@@ -85,21 +85,17 @@ public FileSystemPermission getPermission() {
...
-    }
+  public PermissionStatus applyUMask(FileSystemPermission umask) {
     FileSystemPermission newFileSystemPermission = mPermission.applyUMask(umask);
     return new PermissionStatus(mUserName, mGroupName, newFileSystemPermission);
   }
 
   /**
    * Gets the Directory default {@link PermissionStatus}. Currently the default dir permission is
-   * 0777.
+   * 0777. (Test only)
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: does this mean this is only used in tests?

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 2]
```diff
@@ -108,34 +104,39 @@ public static PermissionStatus getDirDefault() {
...
+  }
 
+  /**
+   * Creates the {@link PermissionStatus} for a file or a directory.
+   *
+   * @param conf the runtime configuration of Alluxio
+   * @return the {@link PermissionStatus} for a file or a directory
+   * @throws java.io.IOException when getting login user fails
+   */
+  public static PermissionStatus getAsLoginUser(Configuration conf) throws IOException {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: fromLoginUser?

so that the invocation looks like `PermissionStatus.fromLoginUser`

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 1 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: yes. This PR is still WIP,  I will probably move it from production to test

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 3]
```diff
@@ -86,52 +90,69 @@ public FileSystemPermission getPermission() {
...
    */
   public PermissionStatus applyUMask(FileSystemPermission umask) {
     FileSystemPermission newFileSystemPermission = mPermission.applyUMask(umask);
     return new PermissionStatus(mUserName, mGroupName, newFileSystemPermission);
   }
 
   /**
-   * Gets the Directory default {@link PermissionStatus}. Currently the default dir permission is
-   * 0777.
+   * Applies default umask to new created files.
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: new -> newly

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 4]
```diff
@@ -86,52 +90,69 @@ public FileSystemPermission getPermission() {
...
    *
-   * @return the default {@link PermissionStatus} for directories
+   * @param conf the runtime configuration of Alluxio
+   * @return a new {@link PermissionStatus}
+   */
+  public PermissionStatus applyFileUMask(Configuration conf) {
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: The method name is bit misleading, because there is no target to apply to.

Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: this method is strange ... I would expect the defaultFilePermissionStatus to be computed by a static factory

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 5]
```diff
@@ -86,52 +90,69 @@ public FileSystemPermission getPermission() {
...
+    FileSystemPermission newFileSystemPermission =
+        mPermission.applyUMask(FileSystemPermission.getUMask(conf)).applyUMask(FILE_UMASK);
+    return new PermissionStatus(mUserName, mGroupName, newFileSystemPermission);
+  }
+
+  /**
+   * Applies default umask to new created directories.
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: ditto

Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: new -> newly

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 6]
```diff
@@ -86,52 +90,69 @@ public FileSystemPermission getPermission() {
...
-        .DEFAULT_FS_FULL_PERMISSION));
+  public PermissionStatus applyDirectoryUMask(Configuration conf) {
+    FileSystemPermission newFileSystemPermission =
+        mPermission.applyUMask(FileSystemPermission.getUMask(conf));
+    return new PermissionStatus(mUserName, mGroupName, newFileSystemPermission);
   }
 
-  // TODO(binfan): remove remote parameter by making two different get methods
   /**
    * Creates the {@link PermissionStatus} for a file or a directory.
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: is this true? it seems like it creates the default permission status for a directory, not a file

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 7]
```diff
@@ -388,6 +388,16 @@ public void setStateTest() throws Exception {
   }
 
   /**
+   * Tests the permission bits are 0755 for directories and 0644 for files by default.
+   */
+  @Test
+  public void permissionTest() throws Exception {
+    mFileSystemMaster.createFile(NESTED_FILE_URI, sNestedFileOptions);
+    Assert.assertEquals(0755, mFileSystemMaster.getFileInfo(NESTED_URI).getPermission());
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: should 0644 and 0755 go into constants?

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java [hunk 8]
```diff
@@ -194,6 +194,14 @@ public void before() throws Exception {
     mPermissionChecker = new PermissionChecker(sTree);
   }
 
+  // Helper function to create a path and set the permission to what specified in option.
+  private static void createAndSetPermission(String path, CreateFileOptions option)
+      throws Exception {
+    InodeTree.CreatePathResult result = sTree.createPath(new AlluxioURI(path), option);
+    result.getCreated().get(result.getCreated().size() - 1)
```
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: why are permissionStatus options not respected in file creation?

Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: I see, they are respected, but then the umask is applied.

**Comment by apc999** [comment]
Jenkins, test this please

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 3 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: done

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 4 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: I think it is because the class members were `final`, not able to alter. I removed the `final` keyword to make code simple.

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 5 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: done

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 6 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: it is true. this method is used by `CreatePathOptions` constructor for both file and directory.

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 7 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: each of these two permissions appeared once here. it will be a bit strange if they are defined here but only get used once.

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: because `sTree.createPath()` applies umask when creating new inodes. Some of the permission bits may be wiped-off after `createPath` even they are set in the options.

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 6 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: how does the constructor decide whether to return 0755 or 0644?

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: I meant across our code base:

```
$ find . -name '*.java' -exec grep -R 0755 {} \; | wc -l => 28
$ find . -name '*.java' -exec grep -R 0644 {} \; | wc -l => 11
```

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: that sounds like a bug

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 6 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: that is not in the constructor. 0755 (for dirs) or 0644 (for files) are created by applying umask for dir and umask for file on top of default permission 0777, the logic to apply umask according to dir or file is in `InodeTree.createPath` which is the entry point for any operations in higher level that add new inode to the existing inode tree.

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 7 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: the reason we didn't put 0755 a constant here is because it is calculated from applying umask for dir on 0777 (this is in a constant). same for 0644.  In other words, they are calculated results (deep in the call stack), rather than being used directly. So I don't really see a clean way to define 0755 or 0644 themselves a constant, maybe only in tests?

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: I should be more specific here. right now we move the logic to apply umask on newly created file or dir on `InodeTree.createPath` which is responsible for creating new inode. That's why by calling `sTree.createPath()`, one could not create some files/dirs with weird permission bits directly (as we are doing here, the tests requires to create a dir with permission bit 0157, in the normal code path, 0157 will be umasked, and the way to create such a file is to update permission later where no umask will be applied )

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 4 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: did you push your change? I still see the members as final

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 6 above)
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: thanks for the explanation

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: thanks for the explanation, I guess my question is: if the permission status option is set by the client, should the umask be still applied? I would expect the umask to only be applied if the permission status option is not set and the default permissions are used

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 7 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: if that's the case, should we be using `PermissionStatus.applyDirUMask(Constant.defaultPermission())` in place of 0755 and `PermissionStatus.applyFileUMask(Constant.defaultPermission())` in place of 0644 everywhere? Except for one place in PermissionStatusTest where we actually check that `PermissionStatus..applyDirUMask(Constant.defaultPermission())` equals 0755 (and similarly for the file umask).

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: as per our offline discussion, I am proposing to get rid of `CreatePathOptions#setPermissionStatus()`

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 4 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: change pushed.  the method is no more a static one, but modifying the member variables.

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 7 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: As discussed offline, we may do this improvement, but not in this PR.

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: Currently I still keep this method because it is now used in (newly added) factory method to create server-side options.

**Comment by apc999** [comment]
@jsimsa  PTAL

I did a significant amount of change actually, mostly regarding how to set default user/group and how to apply umask to permission bits:
- for default `PermissionStatus` it is user:"" group:"" and 777 as permission bits.
- On service handler, when converting thrift wired options to server options, we also set user/group together with  convention, see [CreateDirectoryOptions#fromTOptions](https://github.com/Alluxio/alluxio/pull/2994/files#diff-74128bc262bed0d4f8d94d1c46895eefR46)  and [CreateFileOptions#fromTOptions](https://github.com/Alluxio/alluxio/pull/2994/files#diff-5323790e60da912b117f9742f7e43d3aR47) 
- default permission bits are 777 and file/dir umasks are applied on factory methods of inodes:
[InodeDirectory#create](https://github.com/Alluxio/alluxio/pull/2994/files#diff-576a533e2a6604049a7e87848b61463cR226)  and [InodeFile#create] (https://github.com/Alluxio/alluxio/pull/2994/files#diff-fa65a6f87703f1358bd8aeca869ccdb3R299). In `InodeTree.createPath` which is the entry point for inode creation, we always call the factory method to get new inode (file or dir).

**Comment by apc999** [comment]
@sundapeng @dongche  feel free to review this PR

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 9]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
+  public PermissionStatus applyDirectoryUMask(Configuration conf) {
+    mPermission =
+        mPermission.applyUMask(FileSystemPermission.getUMask(conf));
+    return this;
   }
 
-  // TODO(binfan): remove remote parameter by making two different get methods
   /**
-   * Creates the {@link PermissionStatus} for a file or a directory.
+   * Sets the user inferred from thrift transport for this {@link PermissionStatus}. If
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Sets the user and group ...

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 10]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
-   * @param remote true if the request is for creating permission from client side, the
-   *               username binding into inode will be gotten from {@code AuthenticatedClientUser
-   *               .get().getName()}.
-   *               If the remote is false, the username binding into inode will be gotten from
-   *               {@link alluxio.security.LoginUser}.
    * @return the {@link PermissionStatus} for a file or a directory
    * @throws java.io.IOException when getting login user fails
    */
-  public static PermissionStatus get(Configuration conf, boolean remote) throws IOException {
+  public PermissionStatus setUserFromThriftClient(Configuration conf) throws IOException {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: setUserAndGroupFromThriftClient

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 11]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
+    if (user == null) {
+      throw new IOException(ExceptionMessage.AUTHORIZED_CLIENT_USER_IS_NULL.getMessage());
     }
+    mUserName = user.getName();
+    mGroupName = CommonUtils.getPrimaryGroupName(conf, user.getName());
+    return this;
+  }
 
+  /**
+   * Creates the {@link PermissionStatus} for a file or a directory. If
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Sets the user and group using the login module.

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 12]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
+   * authentication is NOSASL, this a no-op.
+   *
+   * @param conf the runtime configuration of Alluxio
+   * @return the {@link PermissionStatus} for a file or a directory
+   * @throws java.io.IOException when getting login user fails
+   */
+  public PermissionStatus setUserFromLoginModule(Configuration conf) throws IOException {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: setUserAndGroupFromLoginModule

File: core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java [hunk 13]
```diff
@@ -45,127 +42,79 @@
...
+    verifyPermissionStatus("user1", "group1", (short) 0755, permissionStatus);
   }
 
   /**
-   * Tests the {@link PermissionStatus#get(Configuration, boolean)} method.
+   * Tests the {@link PermissionStatus#defaults()} method.
    */
   @Test
-  public void getPermissionStatusTest() throws Exception {
+  public void getDefaultTest() throws Exception {
```
Comment (jsimsa) [inline core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java]: nit: defaultsTest

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 14]
```diff
@@ -296,7 +296,9 @@ public void start(boolean isLeader) throws IOException {
       // Only initialize root when isLeader because when initializing root, BlockMaster needs to
       // write journal entry, if it is not leader, BlockMaster won't have a writable journal.
       // If it is standby, it should be able to load the inode tree from leader's checkpoint.
-      mInodeTree.initializeRoot(PermissionStatus.get(MasterContext.getConf(), false));
+      mInodeTree.initializeRoot(PermissionStatus.defaults()
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: nice, I like this style very much!

File: core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java [hunk 15]
```diff
@@ -212,6 +214,27 @@ public static InodeDirectory fromJournalEntry(InodeDirectoryEntry entry) {
...
+  /**
+   * Creates an {@link InodeDirectory}.
+   *
+   * @param id id of this inode
+   * @param parentId id of the parent of this inode
+   * @param name name of this inode
+   * @param directoryOptions options to create this directory
+   * @return the {@link InodeDirectory} representation
+   */
+  public static InodeDirectory create(long id, long parentId, String name,
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: should we change the constructor to private to enforce that all objects are created using this factory?

File: core/server/src/main/java/alluxio/master/file/meta/InodeFile.java [hunk 16]
```diff
@@ -296,6 +287,31 @@ public static InodeFile fromJournalEntry(InodeFileEntry entry) {
...
+  /**
+   * Creates an {@link InodeFile}.
+   *
+   * @param id id of this inode
+   * @param parentId id of the parent of this inode
+   * @param name name of this inode
+   * @param fileOptions options to create this file
+   * @return the {@link InodeFile} representation
+   */
+  public static InodeFile create(long id, long parentId, String name,
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/meta/InodeFile.java]: should we change the constructor to private to enforce that all objects are created using this factory?

**Comment by jsimsa** [comment]
@apc999 thanks for the improvement, I left some comments

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: I find the problem is that the PermissionStatus object bundles two things that are to be treated differently: 1) user name and group name, which need to be set based on different context (e.g. login module or thrift transport) and 2) permission bits, which do not need to be set

The only way I see fixing this is to get rid of PermissionStatus on the code path that creates and Inode and only (if at all) use it to describe the permission status of an inode. In other words, the CreatePathOptions would have user and group fields (but not permission bits). When inode is created, the user and group is taken from the options and the permission bits are constructed using the defaults + umask.

Comment (jsimsa) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: if you'd prefer, we can leave this change to another PR though

File: core/server/src/main/java/alluxio/master/file/options/CreateDirectoryOptions.java [hunk 17]
```diff
@@ -34,15 +36,22 @@ public static CreateDirectoryOptions defaults() throws IOException {
...
+   * @param tOptions the {@link CreateDirectoryTOptions} to use
+   * @return the {@link CreateDirectoryOptions} from {@link CreateDirectoryTOptions}
    * @throws IOException if an I/O error occurs
    */
-  public CreateDirectoryOptions(CreateDirectoryTOptions options) throws IOException {
-    super();
-    mAllowExists = options.isAllowExists();
-    mPersisted = options.isPersisted();
-    mRecursive = options.isRecursive();
+  public static CreateDirectoryOptions fromTOptions(CreateDirectoryTOptions tOptions)
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/options/CreateDirectoryOptions.java]: why is this no longer a constructor?

File: core/server/src/main/java/alluxio/master/file/options/CreateFileOptions.java [hunk 18]
```diff
@@ -36,17 +37,22 @@ public static CreateFileOptions defaults() throws IOException {
...
+   * @return the {@link CreateFileOptions} from {@link CreateFileTOptions}
    * @throws IOException if an I/O error occurs
    */
-  public CreateFileOptions(CreateFileTOptions options) throws IOException {
-    super();
-    mBlockSizeBytes = options.getBlockSizeBytes();
-    mPersisted = options.isPersisted();
-    mRecursive = options.isRecursive();
-    mTtl = options.getTtl();
+  public static CreateFileOptions fromTOptions(CreateFileTOptions tOptions) throws IOException {
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/master/file/options/CreateFileOptions.java]: why is this no longer a constructor?

File: core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java (hunk 8 above)
Comment (apc999) [inline core/server/src/test/java/alluxio/master/file/PermissionCheckerTest.java]: Yes, I think you get the point --- basically user/group and permission bits are filled/used in different code path. I am cool with you to take on the improvement in followup PRs, and happy to help out.

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 9 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: fixed

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 11 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: thanks for the catch. fixed

File: core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java (hunk 13 above)
Comment (apc999) [inline core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java]: fixed

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 14 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: ack

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 19]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
    *
    * @param umask the umask to apply
    * @return a new {@link PermissionStatus}
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: The doc is outdated.

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 20]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
   }
 
   /**
-   * Gets the Directory default {@link PermissionStatus}. Currently the default dir permission is
-   * 0777.
+   * Applies default umask to newly created files.
    *
-   * @return the default {@link PermissionStatus} for directories
+   * @param conf the runtime configuration of Alluxio
+   * @return a new {@link PermissionStatus}
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: fix javadoc

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 21]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
+    mPermission =
+        mPermission.applyUMask(FileSystemPermission.getUMask(conf)).applyUMask(FILE_UMASK);
+    return this;
+  }
+
+  /**
+   * Applies default umask to newly created directories.
+   *
+   * @param conf the runtime configuration of Alluxio
+   * @return a new {@link PermissionStatus}
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: fix javadoc

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 22]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
-      User user = AuthenticatedClientUser.get(conf);
-      if (user == null) {
-        throw new IOException(ExceptionMessage.AUTHORIZED_CLIENT_USER_IS_NULL.getMessage());
-      }
-      return new PermissionStatus(user.getName(), CommonUtils.getPrimaryGroupName(conf,
-          user.getName()), FileSystemPermission.getDefault().applyUMask(conf));
+    // get the username through the authentication mechanism
+    User user = AuthenticatedClientUser.get(conf);
+    if (user == null) {
+      throw new IOException(ExceptionMessage.AUTHORIZED_CLIENT_USER_IS_NULL.getMessage());
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Why IO exception, but not `AccessControlException `?

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java [hunk 23]
```diff
@@ -86,57 +99,88 @@ public FileSystemPermission getPermission() {
...
+    mGroupName = CommonUtils.getPrimaryGroupName(conf, user.getName());
+    return this;
+  }
 
+  /**
+   * Creates the {@link PermissionStatus} for a file or a directory. If
+   * authentication is NOSASL, this a no-op.
+   *
+   * @param conf the runtime configuration of Alluxio
+   * @return the {@link PermissionStatus} for a file or a directory
```
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Make it clear that it modifies the instance in place.

File: core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java [hunk 24]
```diff
@@ -45,127 +42,79 @@
...
-    permissionStatus = PermissionStatus.get(conf, false);
-    verifyPermissionStatus("test_login_user", "group1", (short) 0755, permissionStatus);
+    // When authentication is enabled, user and group are inferred from thrift transport
+    permissionStatus.setUserFromThriftClient(conf);
+    verifyPermissionStatus("test_client_user", "test_client_user", (short) 0777, permissionStatus);
   }
 
   /**
-   * Tests that retrieving the {@link PermissionStatus} with empty group works as expected.
+   * Tests the {@link PermissionStatus#setUserFromThriftClient(Configuration)} method.
```
Comment (yupeng9) [inline core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java]: fix javadoc

File: core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java (hunk 15 above)
Comment (yupeng9) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: +1

Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java]: done

File: core/server/src/main/java/alluxio/master/file/meta/InodeFile.java (hunk 16 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/meta/InodeFile.java]: done

**Comment by yupeng9** [comment]
Thanks for the improvement. I left some comments.

File: core/server/src/main/java/alluxio/master/file/options/CreateDirectoryOptions.java (hunk 17 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/options/CreateDirectoryOptions.java]: reverted

File: core/server/src/main/java/alluxio/master/file/options/CreateFileOptions.java (hunk 18 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/file/options/CreateFileOptions.java]: reverted

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 19 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Fixed

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 20 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Fixed

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 21 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: Fixed

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 22 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: I didn't change the exception signature but it is left from previous code.

I think both might work in this case. The `AuthenticatedClientUser.get` tries to obtain the user info from IO (here the thrift transport), it also make sense as an IO exception I guess (we haven't really get to security land in this case).

In addition, this signature change will enlarge the footprint of this PR a lot. Do you feel strong on the exception ?

File: core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java (hunk 24 above)
Comment (apc999) [inline core/common/src/test/java/alluxio/security/authorization/PermissionStatusTest.java]: thanks for the catch , fixed

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 23 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: new javadoc updated.

**Comment by apc999** [comment]
@jsimsa @yupeng9 , thanks for the valuable comments and I addressed those comments.
 PTAL.

**Comment by jsimsa** [comment]
LGTM

File: core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java (hunk 22 above)
Comment (yupeng9) [inline core/common/src/main/java/alluxio/security/authorization/PermissionStatus.java]: I dont feel strongly, but I prefer a consistent handling.

**Comment by yupeng9** [comment]
LGTM
